File: code/packages/elixir/mosaic_analyzer/test/mosaic_analyzer_test.exs
defmodule CodingAdventures.MosaicAnalyzerTest do
  use ExUnit.Case, async: true

  alias CodingAdventures.MosaicAnalyzer
  alias CodingAdventures.MosaicAnalyzer.{MosaicComponent, MosaicSlot, MosaicNode, MosaicImport}

  # ===========================================================================
  # Helper — parse and assert success
  # ===========================================================================

  defp ok!(source) do
    case MosaicAnalyzer.analyze(source) do
      {:ok, component} -> component
      {:error, msg} -> flunk("Expected {:ok, component}, got {:error, #{inspect(msg)}}")
    end
  end

  # ===========================================================================
  # 1. Module loads
  # ===========================================================================

  test "module loads" do
    assert Code.ensure_loaded?(CodingAdventures.MosaicAnalyzer)
  end

  # ===========================================================================
  # 2. Minimal component
  # ===========================================================================

  test "analyze/1 — minimal component with no slots" do
    # The simplest possible component: just a Box with no slots or properties.
    component = ok!("component Foo { Box { } }")

    assert %MosaicComponent{} = component
    assert component.name == "Foo"
    assert component.slots == []
    assert component.imports == []
    assert %MosaicNode{} = component.root
    assert component.root.node_type == "Box"
    assert component.root.is_primitive == true
    assert component.root.properties == []
    assert component.root.children == []
  end

  test "analyze/1 — returns {:ok, component} tuple" do
    result = MosaicAnalyzer.analyze("component Bar { Text { } }")
    assert {:ok, %MosaicComponent{name: "Bar"}} = result
  end

  test "analyze/1 — error on syntactically invalid source" do
    # Missing closing brace — parser should fail.
    result = MosaicAnalyzer.analyze("component Foo { Box {")
    assert {:error, _msg} = result
  end

  # ===========================================================================
  # 3. Slot declarations — primitive types
  # ===========================================================================

  test "slot declaration — text type" do
    component = ok!("""
    component Card {
      slot title: text;
      Box { }
    }
    """)

    assert [%MosaicSlot{name: "title", type: {:primitive, "text"}, required: true}] =
             component.slots
  end

  test "slot declaration — number type" do
    component = ok!("""
    component Counter {
      slot count: number;
      Box { }
    }
    """)

    assert [%MosaicSlot{name: "count", type: {:primitive, "number"}, required: true}] =
             component.slots
  end

  test "slot declaration — bool type" do
    component = ok!("""
    component Toggle {
      slot visible: bool;
      Box { }
    }
    """)

    assert [%MosaicSlot{name: "visible", type: {:primitive, "bool"}, required: true}] =
             component.slots
  end

  test "slot declaration — image type" do
    component = ok!("""
    component Avatar {
      slot avatar-url: image;
      Box { }
    }
    """)

    assert [%MosaicSlot{name: "avatar-url", type: {:primitive, "image"}, required: true}] =
             component.slots
  end

  test "slot declaration — color type" do
    component = ok!("""
    component Badge {
      slot bg-color: color;
      Box { }
    }
    """)

    assert [%MosaicSlot{name: "bg-color", type: {:primitive, "color"}, required: true}] =
             component.slots
  end

  test "slot declaration — node type (flexible)" do
    component = ok!("""
    component Wrapper {
      slot content: node;
      Box { }
    }
    """)

    assert [%MosaicSlot{name: "content", type: {:primitive, "node"}, required: true}] =
             component.slots
  end

  # ===========================================================================
  # 4. List slot types
  # ===========================================================================

  test "slot declaration — list<text> type" do
    component = ok!("""
    component List {
      slot items: list<text>;
      Column { }
    }
    """)

    assert [%MosaicSlot{name: "items", type: {:list, {:primitive, "text"}}, required: true}] =
             component.slots
  end

  test "slot declaration — list<node> type" do
    component = ok!("""
    component Shelf {
      slot cards: list<node>;
      Row { }
    }
    """)

    assert [%MosaicSlot{name: "cards", type: {:list, {:primitive, "node"}}, required: true}] =
             component.slots
  end

  # ===========================================================================
  # 5. Component slot type (from imports)
  # ===========================================================================

  test "slot declaration — component type" do
    component = ok!("""
    component Page {
      slot action: Button;
      Box { }
    }
    """)

    assert [%MosaicSlot{name: "action", type: {:component, "Button"}, required: true}] =
             component.slots
  end

  # ===========================================================================
  # 6. Slot default values
  # ===========================================================================

  test "slot with number default value" do
    component = ok!("""
    component Counter {
      slot count: number = 0;
      Box { }
    }
    """)

    [slot] = component.slots
    assert slot.required == false
    assert slot.default_value == {:number, 0.0}
  end

  test "slot with bool default value — true" do
    component = ok!("""
    component Toggle {
      slot visible: bool = true;
      Box { }
    }
    """)

    [slot] = component.slots
    assert slot.required == false
    assert slot.default_value == {:bool, true}
  end

  test "slot with bool default value — false" do
    component = ok!("""
    component Toggle {
      slot visible: bool = false;
      Box { }
    }
    """)

    [slot] = component.slots
    assert slot.default_value == {:bool, false}
  end

  # ===========================================================================
  # 7. Property assignments
  # ===========================================================================

  test "property assignment — dimension value" do
    component = ok!("""
    component Padded {
      Box {
        padding: 16dp;
      }
    }
    """)

    [prop] = component.root.properties
    assert prop.name == "padding"
    assert prop.value == {:dimension, 16.0, "dp"}
  end

  test "property assignment — slot_ref value" do
    component = ok!("""
    component Label {
      slot title: text;
      Text {
        content: @title;
      }
    }
    """)

    [prop] = component.root.properties
    assert prop.name == "content"
    assert prop.value == {:slot_ref, "title"}
  end

  test "property assignment — hex color value" do
    component = ok!("""
    component Colored {
      Box {
        background: #2563eb;
      }
    }
    """)

    [prop] = component.root.properties
    assert prop.name == "background"
    assert prop.value == {:color_hex, "#2563eb"}
  end

  test "property assignment — ident value (bare keyword)" do
    component = ok!("""
    component Aligned {
      Box {
        align: center;
      }
    }
    """)

    [prop] = component.root.properties
    assert prop.name == "align"
    assert prop.value == {:ident, "center"}
  end

  test "property assignment — number value" do
    component = ok!("""
    component Opaque {
      Box {
        opacity: 0.5;
      }
    }
    """)

    [prop] = component.root.properties
    assert prop.name == "opacity"
    assert prop.value == {:number, 0.5}
  end

  # ===========================================================================
  # 8. Nested nodes
  # ===========================================================================

  test "nested node children" do
    component = ok!("""
    component Layout {
      Column {
        Row { }
        Text { }
      }
    }
    """)

    root = component.root
    assert root.node_type == "Column"
    assert length(root.children) == 2
    [{:node, row}, {:node, text}] = root.children
    assert row.node_type == "Row"
    assert text.node_type == "Text"
  end

  # ===========================================================================
  # 9. When blocks
  # ===========================================================================

  test "when block in node children" do
    component = ok!("""
    component Conditional {
      slot show-header: bool;
      Column {
        when @show-header {
          Text { }
        }
      }
    }
    """)

    [{:when_block, slot_name, block_children}] = component.root.children
    assert slot_name == "show-header"
    assert [{:node, %MosaicNode{node_type: "Text"}}] = block_children
  end

  # ===========================================================================
  # 10. Each blocks
  # ===========================================================================

  test "each block in node children" do
    component = ok!("""
    component ItemList {
      slot items: list<text>;
      Column {
        each @items as item {
          Row { }
        }
      }
    }
    """)

    [{:each_block, slot_name, item_name, block_children}] = component.root.children
    assert slot_name == "items"
    assert item_name == "item"
    assert [{:node, %MosaicNode{node_type: "Row"}}] = block_children
  end

  # ===========================================================================
  # 11. Slot references as children
  # ===========================================================================

  test "slot reference used as child" do
    component = ok!("""
    component Page {
      slot header: node;
      Column {
        @header;
      }
    }
    """)

    [{:slot_ref_child, slot_name}] = component.root.children
    assert slot_name == "header"
  end

  # ===========================================================================
  # 12. Import collection
  # ===========================================================================

  test "import declaration is collected" do
    # A non-primitive node type (Button) should appear in imports.
    component = ok!("""
    import Button from "./button.mosaic";
    component Page {
      Box { }
    }
    """)

    assert [%MosaicImport{name: "Button"}] = component.imports
  end

  test "import with alias uses alias as name" do
    component = ok!("""
    import Card as InfoCard from "./card.mosaic";
    component Page {
      Box { }
    }
    """)

    assert [%MosaicImport{name: "InfoCard"}] = component.imports
  end

  # ===========================================================================
  # 13. is_primitive flag on nodes
  # ===========================================================================

  test "primitive nodes have is_primitive true" do
    # These are all the standard primitive nodes.
    primitives = ["Row", "Column", "Box", "Stack", "Text", "Image", "Icon", "Spacer", "Divider", "Scroll"]

    for tag <- primitives do
      component = ok!("component Test { #{tag} { } }")
      assert component.root.is_primitive == true,
             "Expected #{tag} to be primitive"
    end
  end

  test "non-primitive node type has is_primitive false" do
    component = ok!("component Page { Button { } }")
    assert component.root.is_primitive == false
  end

  # ===========================================================================
  # 14. Multiple slots
  # ===========================================================================

  test "multiple slot declarations are collected in order" do
    component = ok!("""
    component Card {
      slot title: text;
      slot subtitle: text;
      slot count: number;
      Box { }
    }
    """)

    assert length(component.slots) == 3
    [s1, s2, s3] = component.slots
    assert s1.name == "title"
    assert s2.name == "subtitle"
    assert s3.name == "count"
  end

  # ===========================================================================
  # 15. analyze_ast/1 with pre-parsed AST
  # ===========================================================================

  test "analyze_ast/1 accepts a pre-parsed file ASTNode" do
    {:ok, ast} = CodingAdventures.MosaicParser.parse("component Foo { Box { } }")
    result = MosaicAnalyzer.analyze_ast(ast)
    assert {:ok, %MosaicComponent{name: "Foo"}} = result
  end

  test "analyze_ast/1 returns error for wrong root rule" do
    fake_ast = %CodingAdventures.Parser.ASTNode{rule_name: "not_a_file", children: []}
    assert {:error, msg} = MosaicAnalyzer.analyze_ast(fake_ast)
    assert msg =~ "file"
  end

  # ===========================================================================
  # 16. Property values — all remaining value kinds
  # ===========================================================================

  test "property assignment — string value" do
    component = ok!("""
    component Named {
      Text {
        label: "Hello";
      }
    }
    """)

    [prop] = component.root.properties
    assert prop.name == "label"
    assert prop.value == {:string, "Hello"}
  end

  test "property assignment — bool true value" do
    component = ok!("""
    component Check {
      Box {
        visible: true;
      }
    }
    """)

    [prop] = component.root.properties
    assert prop.value == {:bool, true}
  end

  test "property assignment — bool false value" do
    component = ok!("""
    component Check {
      Box {
        visible: false;
      }
    }
    """)

    [prop] = component.root.properties
    assert prop.value == {:bool, false}
  end

  test "property assignment — enum value (namespace.member)" do
    component = ok!("""
    component Aligned {
      Box {
        style: heading.small;
      }
    }
    """)

    [prop] = component.root.properties
    assert prop.name == "style"
    assert prop.value == {:enum_val, "heading", "small"}
  end

  # ===========================================================================
  # 17. Default values — additional types
  # ===========================================================================

  test "slot with string default value" do
    component = ok!("""
    component Welcome {
      slot greeting: text = "Hello, World!";
      Text { content: @greeting; }
    }
    """)

    [slot] = component.slots
    assert slot.required == false
    assert slot.default_value == {:string, "Hello, World!"}
  end

  test "slot with color_hex default value" do
    component = ok!("""
    component Themed {
      slot accent: color = #ff5733;
      Box { }
    }
    """)

    [slot] = component.slots
    assert slot.required == false
    assert slot.default_value == {:color_hex, "#ff5733"}
  end

  test "slot with dimension default value" do
    component = ok!("""
    component Spaced {
      slot gap: number = 8;
      Box { }
    }
    """)

    [slot] = component.slots
    assert slot.required == false
    assert slot.default_value == {:number, 8.0}
  end

  # ===========================================================================
  # 18. Multiple children in when/each blocks
  # ===========================================================================

  test "when block with multiple child nodes" do
    component = ok!("""
    component MultiWhen {
      slot show: bool;
      Column {
        when @show {
          Text { }
          Row { }
        }
      }
    }
    """)

    [{:when_block, "show", children}] = component.root.children
    assert length(children) == 2
    [{:node, %MosaicNode{node_type: "Text"}}, {:node, %MosaicNode{node_type: "Row"}}] = children
  end

  test "each block with nested node content" do
    component = ok!("""
    component DetailList {
      slot entries: list<text>;
      Column {
        each @entries as entry {
          Row {
            Text { content: @entry; }
          }
        }
      }
    }
    """)

    [{:each_block, "entries", "entry", [{:node, row}]}] = component.root.children
    assert row.node_type == "Row"
    [{:node, text}] = row.children
    assert text.node_type == "Text"
  end

  # ===========================================================================
  # 19. Scroll, Stack, Icon, Spacer, Divider — full primitive list
  # ===========================================================================

  test "Scroll is a primitive node" do
    component = ok!("component S { Scroll { } }")
    assert component.root.is_primitive == true
  end

  test "Stack is a primitive node" do
    component = ok!("component S { Stack { } }")
    assert component.root.is_primitive == true
  end

  test "Spacer is a primitive node" do
    component = ok!("component S { Spacer { } }")
    assert component.root.is_primitive == true
  end

  test "Divider is a primitive node" do
    component = ok!("component S { Divider { } }")
    assert component.root.is_primitive == true
  end

  test "Icon is a primitive node" do
    component = ok!("component S { Icon { } }")
    assert component.root.is_primitive == true
  end

  # ===========================================================================
  # 20. Property with keyword as property name (e.g., "color: ...")
  # ===========================================================================

  test "property with keyword as property name" do
    # 'color' is a KEYWORD but is also valid as a property name (e.g., SVG color attr)
    component = ok!("""
    component Colorful {
      Box {
        color: #abc123;
      }
    }
    """)

    [prop] = component.root.properties
    assert prop.name == "color"
    assert prop.value == {:color_hex, "#abc123"}
  end

  # ===========================================================================
  # 21. list<node> slots work end to end
  # ===========================================================================

  test "list<node> slot type with each block" do
    component = ok!("""
    component NodeList {
      slot items: list<node>;
      Column {
        each @items as item {
          Box { }
        }
      }
    }
    """)

    [slot] = component.slots
    assert slot.type == {:list, {:primitive, "node"}}
    [{:each_block, "items", "item", _}] = component.root.children
  end
end

File: code/packages/go/mosaic-analyzer/mosaic_analyzer_test.go
// Tests for the Mosaic analyzer.
//
// The analyzer converts a parsed Mosaic AST into a typed MosaicIR,
// resolving types, normalizing values, and identifying primitives.
package mosaicanalyzer

import (
	"testing"
)

// mustAnalyze is a helper that calls Analyze and fatalf on error.
func mustAnalyze(t *testing.T, source string) *MosaicIR {
	t.Helper()
	ir, err := Analyze(source)
	if err != nil {
		t.Fatalf("Analyze error: %v", err)
	}
	if ir == nil {
		t.Fatal("Analyze returned nil")
	}
	return ir
}

// =============================================================================
// TestAnalyzeComponentName
// =============================================================================
//
// The component name should be extracted correctly from the component_decl.
func TestAnalyzeComponentName(t *testing.T) {
	ir := mustAnalyze(t, `component ProfileCard { Column {} }`)
	if ir.Component.Name != "ProfileCard" {
		t.Errorf("Expected name 'ProfileCard', got %q", ir.Component.Name)
	}
}

// =============================================================================
// TestAnalyzeTextSlot
// =============================================================================
//
// A slot with type 'text' should produce Kind="text" in the IR.
func TestAnalyzeTextSlot(t *testing.T) {
	ir := mustAnalyze(t, `
component Label {
  slot title: text;
  Text {}
}`)
	if len(ir.Component.Slots) != 1 {
		t.Fatalf("Expected 1 slot, got %d", len(ir.Component.Slots))
	}
	slot := ir.Component.Slots[0]
	if slot.Name != "title" {
		t.Errorf("Expected slot name 'title', got %q", slot.Name)
	}
	if slot.Type.Kind != "text" {
		t.Errorf("Expected type kind 'text', got %q", slot.Type.Kind)
	}
	if !slot.Required {
		t.Error("Expected slot to be required (no default)")
	}
}

// =============================================================================
// TestAnalyzeAllPrimitiveTypes
// =============================================================================
//
// Each primitive type keyword should map to the correct MosaicType.
func TestAnalyzeAllPrimitiveTypes(t *testing.T) {
	cases := []struct {
		typeName string
		expected string
	}{
		{"text", "text"},
		{"number", "number"},
		{"bool", "bool"},
		{"image", "image"},
		{"color", "color"},
		{"node", "node"},
	}
	for _, tc := range cases {
		source := `component Foo { slot s: ` + tc.typeName + `; Column {} }`
		ir := mustAnalyze(t, source)
		if len(ir.Component.Slots) != 1 {
			t.Errorf("Type %q: expected 1 slot, got %d", tc.typeName, len(ir.Component.Slots))
			continue
		}
		if ir.Component.Slots[0].Type.Kind != tc.expected {
			t.Errorf("Type %q: expected kind %q, got %q", tc.typeName, tc.expected, ir.Component.Slots[0].Type.Kind)
		}
	}
}

// =============================================================================
// TestAnalyzeListType
// =============================================================================
//
// A slot with list<text> type should produce Kind="list" with Element.Kind="text".
func TestAnalyzeListType(t *testing.T) {
	ir := mustAnalyze(t, `
component List {
  slot items: list<text>;
  Column {}
}`)
	if len(ir.Component.Slots) != 1 {
		t.Fatalf("Expected 1 slot, got %d", len(ir.Component.Slots))
	}
	slot := ir.Component.Slots[0]
	if slot.Type.Kind != "list" {
		t.Errorf("Expected kind 'list', got %q", slot.Type.Kind)
	}
	if slot.Type.Element == nil {
		t.Fatal("Expected Element to be set for list type")
	}
	if slot.Type.Element.Kind != "text" {
		t.Errorf("Expected element kind 'text', got %q", slot.Type.Element.Kind)
	}
}

// =============================================================================
// TestAnalyzeSlotWithDefault
// =============================================================================
//
// A slot with a default value should have Required=false and DefaultValue set.
func TestAnalyzeSlotWithDefault(t *testing.T) {
	ir := mustAnalyze(t, `
component Counter {
  slot count: number = 0;
  Text {}
}`)
	slot := ir.Component.Slots[0]
	if slot.Required {
		t.Error("Expected slot to be optional (has default)")
	}
	if slot.DefaultValue == nil {
		t.Fatal("Expected DefaultValue to be set")
	}
	if slot.DefaultValue.Kind != "number" {
		t.Errorf("Expected default kind 'number', got %q", slot.DefaultValue.Kind)
	}
	if slot.DefaultValue.NumValue != 0 {
		t.Errorf("Expected default value 0, got %v", slot.DefaultValue.NumValue)
	}
}

// =============================================================================
// TestAnalyzeBoolDefault
// =============================================================================
//
// A slot with a bool default value.
func TestAnalyzeBoolDefault(t *testing.T) {
	ir := mustAnalyze(t, `
component Conditional {
  slot visible: bool = true;
  Column {}
}`)
	slot := ir.Component.Slots[0]
	if slot.DefaultValue == nil {
		t.Fatal("Expected DefaultValue")
	}
	if slot.DefaultValue.Kind != "bool" {
		t.Errorf("Expected 'bool', got %q", slot.DefaultValue.Kind)
	}
	if !slot.DefaultValue.BoolValue {
		t.Error("Expected default BoolValue=true")
	}
}

// =============================================================================
// TestAnalyzePrimitiveNode
// =============================================================================
//
// Built-in layout nodes (Row, Column, Text, etc.) should have IsPrimitive=true.
func TestAnalyzePrimitiveNode(t *testing.T) {
	primitives := []string{"Row", "Column", "Box", "Stack", "Text", "Image", "Icon", "Spacer", "Divider", "Scroll"}
	for _, tag := range primitives {
		source := `component Foo { ` + tag + ` {} }`
		ir := mustAnalyze(t, source)
		if !ir.Component.Tree.IsPrimitive {
			t.Errorf("Tag %q: expected IsPrimitive=true", tag)
		}
	}
}

// =============================================================================
// TestAnalyzeNonPrimitiveNode
// =============================================================================
//
// Custom/imported component nodes should have IsPrimitive=false.
func TestAnalyzeNonPrimitiveNode(t *testing.T) {
	ir := mustAnalyze(t, `component Card { Button {} }`)
	if ir.Component.Tree.IsPrimitive {
		t.Error("Expected IsPrimitive=false for imported component 'Button'")
	}
	if ir.Component.Tree.Tag != "Button" {
		t.Errorf("Expected tag 'Button', got %q", ir.Component.Tree.Tag)
	}
}

// =============================================================================
// TestAnalyzePropertyDimension
// =============================================================================
//
// A dimension property like `padding: 16dp;` should produce Kind="dimension".
func TestAnalyzePropertyDimension(t *testing.T) {
	ir := mustAnalyze(t, `
component Foo {
  Column {
    padding: 16dp;
  }
}`)
	props := ir.Component.Tree.Properties
	if len(props) != 1 {
		t.Fatalf("Expected 1 property, got %d", len(props))
	}
	prop := props[0]
	if prop.Name != "padding" {
		t.Errorf("Expected 'padding', got %q", prop.Name)
	}
	if prop.Value.Kind != "dimension" {
		t.Errorf("Expected 'dimension', got %q", prop.Value.Kind)
	}
	if prop.Value.NumValue != 16 {
		t.Errorf("Expected NumValue=16, got %v", prop.Value.NumValue)
	}
	if prop.Value.Unit != "dp" {
		t.Errorf("Expected Unit='dp', got %q", prop.Value.Unit)
	}
}

// =============================================================================
// TestAnalyzePropertyColorHex
// =============================================================================
//
// A color property like `background: #2563eb;` should produce Kind="color_hex".
func TestAnalyzePropertyColorHex(t *testing.T) {
	ir := mustAnalyze(t, `
component Foo {
  Column {
    background: #2563eb;
  }
}`)
	props := ir.Component.Tree.Properties
	if len(props) != 1 {
		t.Fatalf("Expected 1 property, got %d", len(props))
	}
	if props[0].Value.Kind != "color_hex" {
		t.Errorf("Expected 'color_hex', got %q", props[0].Value.Kind)
	}
	if props[0].Value.StrValue != "#2563eb" {
		t.Errorf("Expected '#2563eb', got %q", props[0].Value.StrValue)
	}
}

// =============================================================================
// TestAnalyzePropertySlotRef
// =============================================================================
//
// A slot reference in a property value like `content: @title;`.
func TestAnalyzePropertySlotRef(t *testing.T) {
	ir := mustAnalyze(t, `
component Label {
  slot title: text;
  Text { content: @title; }
}`)
	props := ir.Component.Tree.Properties
	if len(props) != 1 {
		t.Fatalf("Expected 1 property, got %d", len(props))
	}
	if props[0].Value.Kind != "slot_ref" {
		t.Errorf("Expected 'slot_ref', got %q", props[0].Value.Kind)
	}
	if props[0].Value.SlotName != "title" {
		t.Errorf("Expected SlotName='title', got %q", props[0].Value.SlotName)
	}
}

// =============================================================================
// TestAnalyzeChildNodes
// =============================================================================
//
// Child nodes in the tree should produce MosaicChild with Kind="node".
func TestAnalyzeChildNodes(t *testing.T) {
	ir := mustAnalyze(t, `
component Card {
  Column {
    Text {}
    Image {}
  }
}`)
	children := ir.Component.Tree.Children
	if len(children) != 2 {
		t.Fatalf("Expected 2 children, got %d", len(children))
	}
	if children[0].Kind != "node" {
		t.Errorf("Expected Kind='node', got %q", children[0].Kind)
	}
	if children[0].Node.Tag != "Text" {
		t.Errorf("Expected tag 'Text', got %q", children[0].Node.Tag)
	}
}

// =============================================================================
// TestAnalyzeSlotRefChild
// =============================================================================
//
// A slot reference used as a child `@action;` produces Kind="slot_ref".
func TestAnalyzeSlotRefChild(t *testing.T) {
	ir := mustAnalyze(t, `
component Card {
  slot action: node;
  Column {
    @action;
  }
}`)
	children := ir.Component.Tree.Children
	if len(children) != 1 {
		t.Fatalf("Expected 1 child, got %d", len(children))
	}
	if children[0].Kind != "slot_ref" {
		t.Errorf("Expected Kind='slot_ref', got %q", children[0].Kind)
	}
	if children[0].SlotName != "action" {
		t.Errorf("Expected SlotName='action', got %q", children[0].SlotName)
	}
}

// =============================================================================
// TestAnalyzeWhenBlock
// =============================================================================
//
// A when block should produce Kind="when" with the correct slot name and body.
func TestAnalyzeWhenBlock(t *testing.T) {
	ir := mustAnalyze(t, `
component Conditional {
  slot show: bool;
  Column {
    when @show {
      Text {}
    }
  }
}`)
	children := ir.Component.Tree.Children
	if len(children) != 1 {
		t.Fatalf("Expected 1 child (when block), got %d", len(children))
	}
	when := children[0]
	if when.Kind != "when" {
		t.Errorf("Expected Kind='when', got %q", when.Kind)
	}
	if when.SlotName != "show" {
		t.Errorf("Expected SlotName='show', got %q", when.SlotName)
	}
	if len(when.Body) != 1 {
		t.Errorf("Expected 1 node in when body, got %d", len(when.Body))
	}
}

// =============================================================================
// TestAnalyzeEachBlock
// =============================================================================
//
// An each block should produce Kind="each" with slot name, item name, and body.
func TestAnalyzeEachBlock(t *testing.T) {
	ir := mustAnalyze(t, `
component List {
  slot items: list<text>;
  Column {
    each @items as item {
      Text {}
    }
  }
}`)
	children := ir.Component.Tree.Children
	if len(children) != 1 {
		t.Fatalf("Expected 1 child (each block), got %d", len(children))
	}
	each := children[0]
	if each.Kind != "each" {
		t.Errorf("Expected Kind='each', got %q", each.Kind)
	}
	if each.SlotName != "items" {
		t.Errorf("Expected SlotName='items', got %q", each.SlotName)
	}
	if each.ItemName != "item" {
		t.Errorf("Expected ItemName='item', got %q", each.ItemName)
	}
}

// =============================================================================
// TestAnalyzeImport
// =============================================================================
//
// Import declarations should appear in the IR's Imports slice.
func TestAnalyzeImport(t *testing.T) {
	ir := mustAnalyze(t, `
import Button from "./button.mosaic";
component Card {
  Column {}
}`)
	if len(ir.Imports) != 1 {
		t.Fatalf("Expected 1 import, got %d", len(ir.Imports))
	}
	imp := ir.Imports[0]
	if imp.Name != "Button" {
		t.Errorf("Expected Name='Button', got %q", imp.Name)
	}
	if imp.Path != "./button.mosaic" {
		t.Errorf("Expected Path='./button.mosaic', got %q", imp.Path)
	}
}

// =============================================================================
// TestAnalyzeMultipleSlots
// =============================================================================
//
// A component with multiple slots should produce them all in order.
func TestAnalyzeMultipleSlots(t *testing.T) {
	ir := mustAnalyze(t, `
component Card {
  slot title: text;
  slot count: number = 0;
  slot visible: bool = true;
  Column {}
}`)
	if len(ir.Component.Slots) != 3 {
		t.Fatalf("Expected 3 slots, got %d", len(ir.Component.Slots))
	}
	if ir.Component.Slots[0].Name != "title" {
		t.Errorf("Slot 0: expected 'title', got %q", ir.Component.Slots[0].Name)
	}
	if ir.Component.Slots[1].Name != "count" {
		t.Errorf("Slot 1: expected 'count', got %q", ir.Component.Slots[1].Name)
	}
	if ir.Component.Slots[2].Name != "visible" {
		t.Errorf("Slot 2: expected 'visible', got %q", ir.Component.Slots[2].Name)
	}
}
